AlphaWallet/Transfer/ViewControllers/SendViewController.swift

```diff
@@ -298,23 +298,36 @@ extension SendViewController: QRCodeReaderDelegate {
         reader.dismiss(animated: true) { [weak self] in
             self?.activateAmountView()
         }
+        guard let result = QRURLParser.from(string: result) else { return }
+        guard checkEIP861Requirements(result: result) else { return }
+    }
 
-        guard let result = QRURLParser.from(string: result) else {
-            return
+    func checkEIP861Requirements(result: ParserResult) -> Bool {
+        //TODO error display on returns
+        //Note: not checking the 'transferType' since erc20 is implied by whether it has uint256 and address.
+        //The contract address can be compared to the one in the token card and if it matches will proceed else fail
+        //this protects the user from sending funds to the wrong address
+        if let chainId = result.params["chainId"] {
+            guard config.chainID == Int(chainId) else { return false }
         }
-        targetAddressTextField.value = result.address
-
-        if let dataString = result.params["data"] {
-            data = Data(hex: dataString.drop0x)
+        //if erc20 (eip861 qr code)
+        if let recipient = result.params["address"], let amt = result.params["uint256"] {
+            amountTextField.ethCost = EtherNumberFormatter.full.string(from: BigInt(amt) ?? BigInt(), units: .ether)
+            targetAddressTextField.value = recipient
+            //address will be set as contract address if erc token, therefore need to ensure the QR code has set the same contract address
+            //as the user is using
+            guard transferType.contract().eip55String == result.address else { return false }
+            return true
         } else {
-            data = Data()
+            targetAddressTextField.value = result.address
         }
-
-        if let value = result.params["amount"] {
-            amountTextField.ethCost = EtherNumberFormatter.full.string(from: BigInt(value) ?? BigInt(), units: .ether)
+        //if ether transfer (eip861 qr code)
+        if let value = result.params["value"], let amountToSend = BigInt(value, radix: 16) {
```

Comment:
Have to handle when it's ERC20, the value isn't decimals=18.

Reply:
Actually, even in the case of erc20 it is 18 decimals, you can artificially lower this but it is up to the qr code maker to get this right. the default is 18 and the units in the QR code are set in whole ethers, meaning for the qr code maker to mess it up they would have to input a value too low to be accepted like 10 decimals points 0.00... when the smart contract accepts 5, however each smart contract still has the 18 decimal point default and the qr maker should be able to get this correct (we cannot know unless we query the smart contract to ask, it might not even be set in the smart contract either). I think it is safe to leave this as is.

Reply:
EIP681 [1] addresses this:

> UNIT is a URL-encoded unicode string. If UNIT is ETH, it always means a multiplier of 10^18. If it is something else AND the addressed contract has a symbol field exactly matching this string AND the contract has a decimals field, then 10 to that power is used as a multiplier. Otherwise, the payment request is deemed invalid. Applications that have no access to the blockchain should refuse accepting requests with a non-empty UNIT, if it is not ETH.

So a correct implementation should fetch `decimals` to verify. https://brunobar79.github.io/eip681-link-generator generates with decimal=0 for ERC20, but yeah, I guess it's not too dangerous to assume decimals=18 since the value sent will be smaller than expected; so we can leave it for now.

[1] https://github.com/ethereum/EIPs/blob/master/EIPS/eip-681.md